### DISCUSSION TITLE
Does PNP Powershell Decorate Traffic?

### DISCUSSION THREAD

**Discussion by Andre-DuToit** [pnp/powershell#2543] [Q&A] [answered]
Hi All,

As most questions on here relate to rate limits and throttling, i was wondering if the cmdlets decorate traffic in order to allow sharepoint to prioritize it.

To that point, does PnP Use CSOM, REST API or Microsoft Graph to establish connectivity and manage the Sharepoint Site? I'm currently struggling to get past rate limiting due to thousands of Sharepoint List items I manage via Powershell. I hit rate limits not long after using the Module, and only recently started using batching (only for archival and recycling the site list items) as set-pnplistitem doesnt like non string values in the batch.

Any answers or ideas as to how i can better connect and get the list items in my sharepoint tenant would be appreciated

**Comment by AndersRask** [comment, marked as answer]
Some of the underlying frameworks handle user agent decoration, like NONISV|SharePointPnP|PnPCore/ in PnPHttpClient
In most traffic the PnPFramework RetryHandler will handle soft throttling retries (429)

But it is traffic just as all other traffic, and PnP.* are not officially supported frameworks, so no prioritation is done to my knowledge.

PnP.PowerShell uses almost exclusively CSOM, so you could consider using Graph that tend to use less Resource Units on your RateLimit (amount is depending on how many licenses you have) and also performs much faster.
If you need to use PnP.PowerShell on large lists, make sure your CAML query uses indexed fields and make sure your first filter is the one that limits requests below the throttle limit

**Reply by Andre-DuToit** [reply]
I'm currently only selecting fields i need with the -Fields value, for exactly the reason you mentioned. So when  you say, use Microsoft Graph, i assume you are talking about the Microsoft.Graph.Sites module?

**Reply by AndersRask** [reply]
How you use Graph is up to your preference and the scenario

https://learn.microsoft.com/en-us/graph/overview

The PowerShell module is definitely an option
https://github.com/microsoftgraph/msgraph-sdk-powershell
https://lazyadmin.nl/powershell/install-microsoft-graph-module/

Practical usage
https://practical365.com/sharepoint-site-files-report/

**Reply by Andre-DuToit** [reply]
Thanks Anders, ill have a look to see what options fit better. Appreciate the time you took to answer my question.

Have a good day!

**Reply by AndersRask** [reply]
Glad to help!
